### ISSUE TITLE
Add new 'resolution' replacement token

### ISSUE THREAD

**Issue by lsrom** [tvrenamer/tvrenamer#114]
Hello. I was wondering if it would be possible to add new token, episode resolution.
For example from TV show `Example.S01E02.720p.HDTV.X264-DIMENSION[ettv]` i would like to have `Example s01e02 [720p]`.
I suspect it is not a change that would take long time to write for any involved developer. If it is, I could write it myself and add it in a pull request but I don't have much time to familiarize myself with new code base and thus can't tell how long before I could submit my code.
Thank you in advance for any response.

**Comment by daveharris** [comment]
Hi @lsrom,

This would be a nice addition as I do that myself for movies so I can see the use case for it.

If you could submit a pull request that would be awesome. As you can probably tell, this project doesn't have a lot of time spent on it as we are so busy.

You could search all the code for `EPISODE_NUM_LEADING_ZERO` for places which you will need to update.

There should be (at least) 4 places:

1. `src/main/com/google/code/tvrenamer/model/ReplacementToken.java` to define the token to replace, maybe `%r` for resolution?

1. `src/main/com/google/code/tvrenamer/view/PreferencesDialog.java` to add it to the Preferences window

1. `src/main/com/google/code/tvrenamer/model/FileEpisode.java` to perform the replacement. The work here will be coming up the a regular expression which extracts the correct information. I suggest you make it simple/dumb and we can refine later if need be. I suggest having a play with something like [Rubular](http://rubular.com/r/9Sq0P8ac55) to get the regex correct first. See [my quick test on Rubular](http://rubular.com/r/9Sq0P8ac55) and come up with the regex `\W(\d+[pk])\W` which will accept "720p", "1080p" or "4k". I have a feeling that the non-whitespace (`\W`) have already been stripped out at that point so you might need to be a bit more tricky... Good luck, have fun!

1. Update the tests to ensure your code a) works, and b) doesn't break anything else

I hope that gives you something to start off with. If you get stuck, let us know.

I'll mark this an an enhancement and leave it open for now.

Dave

**Comment by lsrom** [comment]
Hi @daveharris,
I started working on it and already have a problem. I have working solution for files containing info about their resolution but it doesn't work for those files without it.  
Let's take for example the first regular expression: `(.+?\d{4}\W\D*?)[sS]?(\d\d?)\D*?(\d\d).*`.
I changed it to: `(.+?\d{4}\W\D*?)[sS]?(\d\d?)\D*?(\d\d).*\D(\d+[pk]).*`.

Now I wanted to wrap the last part I added in `?` so it can match those file names without the resolution set. But that completely changes the results.  
With added `?` the xpression would be: `(.+?\d{4}\W\D*?)[sS]?(\d\d?)\D*?(\d\d)(.*\D(\d+[pk]))?.*`.
You can check it out [here](http://rubular.com/r/EiGayHPnQu).  
I am not able to find a solution to this so far so I would like to ask if there is obvious mistake I am not seeing or if I should start looking for different solution.  
One solution which doesn't require any regex magic is simply to use both previous regular expressions and new ones with resolution.  
Sorry for bothering with such a newbie question, I know your time is limited.

**Comment by vipuldelwadia** [comment]
What we do in tvrenamer is have a set of regexes, and try them one after another until we have a match, so ideally your regex needs to match if the resolution piece is there, and fail if it isn't.

What you don't want is your regex (incorrectly) matching on inputs without resolutions and getting the tokens wrong.

**Comment by lsrom** [comment]
Hi @vipuldelwadia,
that much I know. It looks like I can't fix the regex on my own, so I asked for help or another solution.  
As for the another solution, I think it would be possible to use all the regexes, old ones and new ones together without the need to fix the regex but I don't want to do that if there is a simple solution to fix the incorrect matches.  
The code I wrote is otherwise working, this is the only thing left to solve before I can make pull request.

**Comment by vipuldelwadia** [comment]
Right, I see what you mean. The issue you've run into is something we've run into many, many times while doing tvrenamer (have a look at the git log!) - basically the regex is too complicated and not understandable, but the bigger problem is the input string we're trying to regex is complicated too.

We probably need to approach the problem differently, perhaps tokenising before extracting information? Or some other sort of heuristic based approach?

**Comment by lsrom** [comment]
Perhaps the simplest thing that can be done is this:

```
public static final String[] REGEX = {
			"(.+?\\d{4}\\W\\D*?)[sS]?(\\d\\d?)\\D*?(\\d\\d).*\\D(\\d+[pk]).*", 
			"(.+?\\W\\D*?)[sS](\\d\\d?)[eE](\\d\\d?).*\\D(\\d+[pk]).*", 
			"(.+\\W\\D*?)[sS](\\d\\d?)\\D*?[eE](\\d\\d).*\\D(\\d+[pk]).*", 
			"(.+\\W\\D*?)(\\d\\d?)\\D+(\\d\\d).*\\D(\\d+[pk]).*", 
			"(.+\\W+)(\\d\\d?)(\\d\\d).*\\D(\\d+[pk]).*"
	};

while (idx < COMPILED_REGEX.length) {
    if (matcher.matches() && matcher.groupCount() == 4) {
        // matching regex with resolution part
    } else {
        regex.replace("\\D(\\d+[pk]).*", "");
        if (matcher.matches() && matcher.groupCount() == 3) {
            // match regex without resolution (as it is now)
        }
    }
}
```

This here is not real code, it's just to illustrate what I have in mind.  Also, I haven't tried it yet, it's just something I thought of.  
Basically, it would match regex containing resolution matching and if it failed it would revert back to current regular expressions by replacing the resolution matching part with "".  
Is this something you would be fine with? It would double the amount of regular expression matches done but I think that is not such a problem. What might be a problem is worse readibility of the code but that can be solved by thorough commenting.

**Comment by vipuldelwadia** [comment]
That's what I was trying to get at in my first comment. Basically I don't mind adding more regexes as long as it doesn't break current behaviour.

There are also tests you should look at running once you've implemented your changes, and you should add your new use cases to the tests to make sure it works.

**Comment by lsrom** [comment]
Just added pull request #115 which implements resolution matching and replacement.  
Check it out and if you have any comments, let me know.

**Comment by eprenamer** [comment]
@lsrom Now fixed in v0.8

https://github.com/tvrenamer/tvrenamer/releases/latest
